### PR TITLE
Untangle: Add Subscribers to Jetpack Cloud (v2)

### PR DESCRIPTION
**Pull request by lupus2k**
Related to https://github.com/Automattic/dotcom-forge/issues/5445

Because of a bad merge, this is the follow-up of https://github.com/Automattic/wp-calypso/pull/87615

## Proposed Changes

Add Subscribers page to Jetpack Cloud.

#### TO-DO
- [x] Import Endpoint responds `Invalid` (Related to fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Serfg%2Qncv%2Qcyhtvaf%2Sraqcbvagf%2Sfhofpevoref.cuc%3Se%3Qpnq3o86s%23360-og )
- [x] Check Migrate from other WordPress sites.
 
## Testing Instructions

* Run yarn start-jetpack-cloud
* Go to http://jetpack.cloud.localhost:3000/
* Go to /subscribers
* Check if everything is working (except TO-DOs items)
* Test it on Jetpack Self hosted & Atomic sites.

### CODE REVIEW

**Review by candy02058912 — APPROVED** [review]
Was able to see the /subscribers and able to import subscribers after applying D139222-code

subscribers@2x

**Comment by miksansegundo** [comment]
Sharing testing notes: 

## Atomic sites

### The subscribers screen mainly works

Screenshot 2567-02-23 at 12 09 22


🐛 There is a bug when there aren't any subscribers because the Launchpad component is not working on Atomic and Jetpack sites. This issue exists in production too. It was introduced in https://github.com/Automattic/wp-calypso/pull/83996

|EXPECTED|RESULT|
|-|-|
|Screenshot 2567-02-23 at 21 21 07|Screenshot 2567-02-23 at 12 10 25|

### The import modal works
Screenshot 2567-02-23 at 11 56 18

The request is POST `/subscribers/import`.

However, the email that the notice says I should receive after the import wasn't delivered.

Screenshot 2567-02-23 at 12 00 34



### The migration tool doesn't work

Screenshot 2567-02-23 at 11 38 30

The request POST `/migrate` returns 405 "Not authorized" because the method is not allowed.

## Jetpack sites

The same results as for Atomic, except that the screen shown when there aren't subscribers is different:

Screenshot 2567-02-23 at 12 39 47

I'll continue investigating to fix these issues. cc: @ebinnion @lsl @taipeicoder

**Comment by miksansegundo** [comment]
I have added some changes to this PR:

- [x] Fixed the migration feature using `calypso/lib/wp` for the request so the user gets authenticated. 
  - Tested on Calypso green and blue with Atomic, Simple, and Jetpack sites.
  - Verified that an email is sent when the migration is completed.
- [x] Updated the icon. Asked Jetpack designers in p1708701853576159-slack-C0D96691V

TODOs:

- [x] Investigate why the Launchpad component, shown when there aren't subscribers, doesn't work on some Atomic sites.

|Atomic 1|Atomic 2|
|-|-|
|Screenshot 2567-02-23 at 23 40 09|Screenshot 2567-02-23 at 23 39 39|

File: client/jetpack-cloud/sections/sidebar-navigation/manage-selected-site.tsx [hunk 1]
```diff
@@ -131,6 +132,15 @@ const useMenuItems = ( {
 					enabled: isAdmin && isSectionNameEnabled( 'jetpack-social' ) && ! isWPForTeamsSite,
 					isSelected: itemLinkMatches( path, `${ JETPACK_CLOUD_SOCIAL_LINK }/${ siteSlug }` ),
 				},
+				{
+					icon: <JetpackIcons icon="subscribers" size={ 24 } />,
```
Comment (simison) [inline client/jetpack-cloud/sections/sidebar-navigation/manage-selected-site.tsx]: Did you try using WP icon directly instead?

```jsx
import { Icon, people } from '@wordpress/icons';

<Icon icon={ people } size={ 24 } />
```

**Comment by simison** [comment]
@Automattic/apex @daledupreez do you know what's up with the launchpad component above https://github.com/Automattic/wp-calypso/pull/87684#issuecomment-1961660982?

**Comment by crisbusquets** [comment]
Hi!

I tested it and found some details. Let me know if I should create issues for them:

JN sites Launchpad behaves differently than the Atomic ones. The correct design is the one from Atomic, including the actions. We shouldn't open Support Pages here, users will end up in a rabbit hole.

Screenshot 2024-02-26 at 13 38 40


The heading font size isn't right, here are all the design specs (spacings and typography): S4xGQTEkLhYdUvcIdn2Cfr-fi-67_16943.


### Turn visitors into subscribers
In JN it opens [support pages](https://wordpress.com/support/wordpress-editor/blocks/subscribe-block/) and on Atomic, a modal.


### Add the subscribe block to your site
JN opens [support pages](https://wordpress.com/support/import-subscribers-to-a-newsletter/), Atomic doesn't work. It should open the site editor, with the Subscribe block: `/site-editor/jetpackgoku.wpcomstaging.com?canvas=edit&help-center=subscribe-block`


### Share your site
JN opens a [support page](https://wordpress.com/support/category/grow-your-audience/) and Atomic the Share modal.

**Comment by daledupreez** [comment]
@simison and/or @miksansegundo, could you check whether the problematic back end Atomic site is running a custom/dev version of `wpcomsh`? That's the most likely cause of the issue, as the Launchpad REST API validates that the checklist slug exists locally:
https://github.com/Automattic/jetpack/blob/253e44209c31006822c2c2d396ad2ebbb4eb165b/projects/packages/jetpack-mu-wpcom/src/features/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-launchpad.php#L42

Side note: `jetpack-mu-wpcom` is loaded as a Composer dependency of `wpcomsh` on Atomic sites.

File: client/jetpack-cloud/sections/sidebar-navigation/manage-selected-site.tsx (hunk 1 above)
Comment (miksansegundo) [reply client/jetpack-cloud/sections/sidebar-navigation/manage-selected-site.tsx]: I did not. Would that be better? I'll update it if that's preferred.

**Comment by miksansegundo** [comment]
> could you check whether the problematic back end Atomic site is running a custom/dev version of wpcomsh?

@daledupreez, that was it. That site is a testing site that was not using the latest version of wpcomsh. Thanks for spotting it.

**Comment by miksansegundo** [comment]
There are still issues only on Calypso green because the requests made from Launchpad are not authenticated. The error is a 401 `rest_forbidden`.

Launchpad requests work when we use `wpcom.req` instead of `wpcomRequest` because the user needs to be authenticated via `Authorization: Bearer` rather than via `Authorization: X-WPCOOKIE`.

|BEFORE|AFTER|
|-|-|
|Screenshot 2567-02-27 at 15 28 49|Screenshot 2567-02-27 at 14 46 49|

The commit https://github.com/Automattic/wp-calypso/pull/87684/commits/1c9f6f8da8d2ffa46c1c49445b803afe5654b700 is an example, that fixes the request to [fetch a Launchpad checklist](https://github.com/Automattic/wp-calypso/blob/7e9c62e8de2a6d69d078f84a6646138e1d238d14/packages/data-stores/src/queries/use-launchpad.ts#L64). However, it's required also to disable the linter with `/* eslint-disable no-restricted-imports */` because the import for `calypso/lib/wp` is restricted there.

We'll have to update all the requests related to the Launchpad for Subscribers to ensure the user is authenticated.

@daledupreez @simison, what are your thoughts on refactoring Launchpad partially to use `wpcom.req`?

**Comment by miksansegundo** [comment]
The commit https://github.com/Automattic/wp-calypso/pull/87684/commits/c12495d62f4eece013a368472bbcab4289b4848c is another example of how we can fix the request authentication issue.

It fixes the issue with the emails not being sent when a subscriber is imported via the following modal:

Screenshot 2567-02-27 at 16 08 48

Users see this notice:
Screenshot 2567-02-27 at 16 05 20

And the email is received:

Screenshot 2567-02-27 at 16 17 07

**Comment by miksansegundo** [comment]
> JN sites Launchpad behaves differently than the Atomic ones. The correct design is the one from Atomic, including the actions. We shouldn't open Support Pages here, users will end up in a rabbit hole.

Noting that currently, in production, we show the Subscribers Launchpad only to Atomic sites. For Jetpack sites, we show the [empty-list-view](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/subscribers/components/empty-list-view/empty-list-view.tsx) component. 


### Production
|Atomic|Jetpack|
|-|-|
|Screenshot 2567-02-28 at 15 19 56|Screenshot 2567-02-28 at 15 20 05|

**Comment by lsl** [comment]
> For Jetpack sites, we show the [empty-list-view](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/subscribers/components/empty-list-view/empty-list-view.tsx) component.

Can we do the same for Atomic sites on jetpack cloud? Leaving the launchpad behavior in dotcom default sites only?

**Comment by autumnfjeld** [comment]
> For Jetpack sites, we show the [empty-list-view](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/subscribers/components/empty-list-view/empty-list-view.tsx) component.

> Can we do the same for Atomic sites on jetpack cloud? Leaving the launchpad behavior in dotcom default sites only?

Is there any functionality loss? If it is only guidance, seems fine to me.

**Comment by miksansegundo** [comment]
>JN opens [support pages](https://wordpress.com/support/import-subscribers-to-a-newsletter/), Atomic doesn't work. It should open the site editor, with the Subscribe block: /site-editor/jetpackgoku.wpcomstaging.com?canvas=edit&help-center=subscribe-block

The task "Add the subscribe block to your site" opens the editor with the param `help-center=subscribe-block` to show the block's help center. See https://github.com/Automattic/wp-calypso/pull/84609

The task link is a Calypso path to open the editor in wordpress.com, which doesn't exist within the Jetpack Manage domain (cloud.jetpack.com). See [launchpad-task-definitions.php#L575](https://github.com/Automattic/jetpack/blob/045bfee670dc19ccda3994ae0d367de3e4564df2/projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php#L575)

I intend to find a way to open the link in wordpress.com dynamically by checking the browser `location.host` because I'm unsure if it is a good idea to use an absolute URL in the task definition. cc: @daledupreez

**Comment by miksansegundo** [comment]
> Is there any functionality loss? If it is only guidance, seems fine to me.

Yes, there is functionality loss because the items in [empty-list-view](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/subscribers/components/empty-list-view/empty-list-view.tsx) are only links to support docs while the other component has actions.

The support docs are:

https://wordpress.com/support/wordpress-editor/blocks/subscribe-block/
https://wordpress.com/support/import-subscribers-to-a-newsletter/
https://wordpress.com/support/category/grow-your-audience/

**Comment by simison** [comment]
> Is there any functionality loss? If it is only guidance, seems fine to me.

Let's not please remove the empty state launchpad for Atomic sites. It's working on Calypso side just fine:

image

What's the problem exactly for it not working with Jetpack Cloud?

The same launchpad is also used for empty subscriber stats (also for Atomic sites):

Screenshot 2024-02-28 at 13 40 03

For self-hosted Jetpack sites fine to keep as-is today without launchpad.

> I intend to find a way to open the link in wordpress.com dynamically by checking the browser location.host because I'm unsure if it is a good idea to use an absolute URL in the task definition. 

Yep, linking to site editor in .com is fine (assuming it'll end up in wp-admin instead of iframed eventually). You can also use selectors to pull the site editor URL, example:

https://github.com/Automattic/wp-calypso/blob/409a670741634cd9298cc31f46bc49413e93ca23/client/my-sites/site-settings/settings-newsletter/SubscribeModalSetting.tsx#L27-L30

We'll be replacing the task to a new flow soonish anyway as the current isn't great, but not fast enough to support this work.

**Comment by miksansegundo** [comment]
Just did a quick test to see if Launchpad would work with Jetpack sites and found that the API returns a 404 `rest_no_route`.


https://github.com/Automattic/wp-calypso/assets/1881481/100b1f5d-7b02-445e-965d-4ed0301061a3

**Comment by miksansegundo** [comment]
> Yep, linking to site editor in .com is fine (assuming it'll end up in wp-admin instead of iframed eventually).

Initially, I solved that in https://github.com/Automattic/wp-calypso/pull/87684/commits/a53a5dc5a09bf83a0946bd0abb62c1bc78ddaa2f but then reverted it because that URL is being updated in Jetpack via https://github.com/Automattic/jetpack/pull/36014 

> For self-hosted Jetpack sites fine to keep as-is today without launchpad.

👍 

> What's the problem exactly for it not working with Jetpack Cloud?

At the moment, all issues seem solved. The three tasks are working:

https://github.com/Automattic/wp-calypso/assets/1881481/9df265fe-d452-4f50-b9d3-f17cbd5a015a

**Review by fushar — COMMENTED** [review]
I tried the Add Subscribers modal, but it shows a warning "You are not allowed to manage subscribers", which I do not see in Calypso. Is this expected?

|This PR|Using Calypso Users -> Subscribers|
|-|-|
|image|image|

The site is Atomic (not WoA), without Jetpack Beta Tester.

EDIT: It looks like I need to apply D139222-code first, would be great if this was added to the test instructions 😄

**Review by fushar — APPROVED** [review]
Tested the following on Atomic site and a JN site (after applying D139222-code and https://github.com/Automattic/jetpack/pull/36014)

✅ There is a new Subscribers page on Jetpack Cloud
✅ The import modal works
✅ Download CSV works
✅ Migrate from other site works
✅ Launchpad tasks work

Nice job 🎉

**Comment by candy02058912** [comment]
Tested with the following roles:
- Editor
- Author
- Contributor
- Subscriber
- Customer
- Shop Manager
They don't have access to /subscribers
Left a note on the weird backup redirects (p1709644225680709/1709620654.548899-slack-C06DN6QQVAQ), which I don't think is a blocker for this PR.

**Review by candy02058912 — COMMENTED** [review]
Tested with the following roles:
- Editor
- Author
- Contributor
- Subscriber
- Customer
- Shop Manager

They don't have access to /subscribers
Left a note on the weird backup redirects (p1709644225680709/1709620654.548899-slack-C06DN6QQVAQ), which I don't think is a blocker for this PR.

**Comment by simison** [comment]
Heads up, these cards links look to be broken at the bottom of the page:

image

Stats link specifically should probably link a bit different places depending on is it WP.com or Jetpack site, and state of calypso-untangling for WP.com?

**Comment by simison** [comment]
Importing is throwing an error for both Jetpack and WP.com sites:

image

The importing itself did work but error is a problem.

Also seeing weird things happen with site switcher:


https://github.com/Automattic/wp-calypso/assets/87168/ef17901e-7554-4862-aade-00c9d3373fdc

**Comment by lupus2k** [comment]
> Also seeing weird things happen with site switcher:

I noticed this also happens on Production and with all menus 🤔

**Comment by arthur791004** [comment]
> Also seeing weird things happen with site switcher:

It seems to be the expected behavior as [we set the base path to /landing for Jetpack Cloud](https://github.com/Automattic/wp-calypso/blob/3b7ed3229b2821dc527625201bf47408179949a1/client/jetpack-cloud/components/sidebar/site-selector.tsx#L58) and [site selector switches to that path and you switch the site](https://github.com/Automattic/wp-calypso/blob/3b7ed3229b2821dc527625201bf47408179949a1/client/components/site-selector/index.jsx#L558).

**Pull request merged**